File: sdks/flutter/test/client_test.dart
/// Tests unitarios del SDK Flutter — 4 invariantes:
///   1. queue cap descarta cuando se llena
///   2. retry on 5xx
///   3. beforeSend filtra (null → descartar)
///   4. scrubbing aplica scrubFields + scrubPatterns
///
/// Levantamos un HttpServer real en localhost (puerto 0 = libre) en cada test;
/// no usamos mocks porque la SDK pasa por package:http → dart:io igualmente.

import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:faro_sdk/faro_sdk.dart';

class _Capture {
  final HttpServer server;
  final List<Map<String, dynamic>> batches = [];
  int calls = 0;
  int nextStatus = 200;

  _Capture._(this.server) {
    server.listen((req) async {
      calls++;
      final raw = await utf8.decoder.bind(req).join();
      try {
        batches.add(jsonDecode(raw) as Map<String, dynamic>);
      } catch (_) {
        batches.add({'_raw': raw});
      }
      req.response.statusCode = nextStatus;
      req.response.headers.contentType = ContentType.json;
      req.response.write('{"ok":true}');
      await req.response.close();
    });
  }

  static Future<_Capture> start() async {
    final s = await HttpServer.bind('127.0.0.1', 0);
    return _Capture._(s);
  }

  Uri get endpoint => Uri.parse('http://127.0.0.1:${server.port}');

  Future<void> close() => server.close(force: true);
}

void main() {
  // ---- 1. queue cap ----
  test('queue cap descarta cuando se llena', () async {
    final cap = await _Capture.start();
    try {
      // maxBatchSize == maxQueueSize → un flush vacía toda la cola.
      // Si el cap se respeta, el server recibe a lo sumo 5 eventos en un único batch.
      final faro = Faro.init(FaroOptions(
        endpoint: cap.endpoint.toString(),
        token: 'tk',
        service: 'queue-cap',
        installGlobalHandlers: false,
        flushInterval: const Duration(seconds: 100), // sin auto-flush
        maxBatchSize: 50,
        maxQueueSize: 5,
      ));
      try {
        for (var i = 0; i < 50; i++) {
          faro.log(level: 'INFO', message: 'evento $i');
        }
        await faro.flush();
        // El primer batch (el único, dada la config) debe tener ≤5 eventos.
        await _waitFor(() => cap.batches.isNotEmpty);
        final logs = (cap.batches.first['logs'] as List).cast<dynamic>();
        expect(logs.length, lessThanOrEqualTo(5),
            reason: 'el cap se aplica en log(), nunca enviamos más');
      } finally {
        await faro.close();
      }
    } finally {
      await cap.close();
    }
  });

  // ---- 2. retry on 5xx ----
  test('5xx: el batch se re-encola', () async {
    final cap = await _Capture.start();
    try {
      cap.nextStatus = 503;
      final faro = Faro.init(FaroOptions(
        endpoint: cap.endpoint.toString(),
        token: 'tk',
        service: 'retry-test',
        installGlobalHandlers: false,
        flushInterval: const Duration(seconds: 100),
      ));
      try {
        faro.log(level: 'INFO', message: 'reintentar-me');
        await faro.flush(); // 503 → el SDK re-encola
        cap.nextStatus = 200;
        await faro.flush(); // ahora OK
        await _waitFor(() => cap.calls >= 2);
        expect(cap.calls, greaterThanOrEqualTo(2),
            reason: 'segundo intento debe haber llegado');
      } finally {
        await faro.close();
      }
    } finally {
      await cap.close();
    }
  });

  // ---- 3. beforeSend ----
  test('beforeSend null descarta', () async {
    final cap = await _Capture.start();
    try {
      final faro = Faro.init(FaroOptions(
        endpoint: cap.endpoint.toString(),
        token: 'tk',
        service: 'bs-discard',
        installGlobalHandlers: false,
        flushInterval: const Duration(seconds: 100),
        beforeSend: (e) => e.message.contains('descartar') ? null : e,
      ));
      try {
        faro.log(level: 'INFO', message: 'guardar');
        faro.log(level: 'INFO', message: 'descartar');
        faro.log(level: 'INFO', message: 'guardar también');
        await faro.flush();
        await _waitFor(() => cap.batches.isNotEmpty);
        final msgs = (cap.batches.first['logs'] as List)
            .cast<Map<String, dynamic>>()
            .map((l) => l['message'] as String)
            .toList();
        expect(msgs, equals(['guardar', 'guardar también']));
      } finally {
        await faro.close();
      }
    } finally {
      await cap.close();
    }
  });

  // ---- 4. scrubbing ----
  test('scrubbing aplica scrubFields y scrubPatterns', () async {
    final cap = await _Capture.start();
    try {
      final faro = Faro.init(FaroOptions(
        endpoint: cap.endpoint.toString(),
        token: 'tk',
        service: 'scrub',
        installGlobalHandlers: false,
        flushInterval: const Duration(seconds: 100),
      ));
      try {
        faro.log(
          level: 'INFO',
          message: 'auth con eyJabc.def.ghi y key sk-abcdefghijklmnop',
          attributes: {
            'user.password': 'p4ssw0rd',
            'http.request.header.authorization': 'Bearer x',
            'safe.field': 'visible',
            'embedded': 'ghp_aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa',
          },
        );
        await faro.flush();
        await _waitFor(() => cap.batches.isNotEmpty);
        final log = (cap.batches.first['logs'] as List).first
            as Map<String, dynamic>;
        final attrs = log['attributes'] as Map<String, dynamic>;
        expect(attrs['user.password'], '[REDACTED]');
        expect(attrs['http.request.header.authorization'], '[REDACTED]');
        expect(attrs['safe.field'], 'visible');
        expect(attrs['embedded'], '[REDACTED]');
        final msg = log['message'] as String;
        expect(msg.contains('eyJabc'), isFalse,
            reason: 'JWT redactado en message');
        expect(msg.contains('sk-abcdef'), isFalse,
            reason: 'sk-* redactado en message');
      } finally {
        await faro.close();
      }
    } finally {
      await cap.close();
    }
  });
}

Future<void> _waitFor(bool Function() cond,
    {Duration timeout = const Duration(seconds: 2)}) async {
  final deadline = DateTime.now().add(timeout);
  while (DateTime.now().isBefore(deadline)) {
    if (cond()) return;
    await Future<void>.delayed(const Duration(milliseconds: 30));
  }
}

File: sdks/expo/test/client.test.mjs
/**
 * Tests unitarios del SDK Expo — 4 invariantes:
 *   1. queue cap
 *   2. retry on 5xx
 *   3. beforeSend filtra
 *   4. scrubbing
 *
 * No requiere react-native ni AsyncStorage: pasamos `installGlobalHandlers: false`
 * y `persistence: false` para evitar los `require('react-native')` /
 * `require('@react-native-async-storage/async-storage')`.
 */

import { test } from 'node:test';
import assert from 'node:assert/strict';
import http from 'node:http';
import { createRequire } from 'node:module';

// El bundle Expo es CJS (main: dist/index.js). Cargamos con createRequire para
// que el `require()` interno del SDK funcione (Metro lo expone en producción).
const faro = createRequire(import.meta.url)('../dist/index.js');

function startServer(handler) {
  return new Promise((resolve) => {
    const server = http.createServer((req, res) => {
      let body = '';
      req.on('data', (c) => (body += c));
      req.on('end', () => handler(req, res, body));
    });
    server.listen(0, () => resolve({ server, port: server.address().port }));
  });
}

const delay = (ms) => new Promise((r) => setTimeout(r, ms));

function commonOpts(port) {
  return {
    endpoint: `http://127.0.0.1:${port}`,
    token: 'tk',
    service: 'expo-tests',
    installGlobalHandlers: false,
    persistence: false,
    flushIntervalMs: 100_000,
  };
}

// ---- 1. queue cap ----

test('queue cap: enqueue silencioso al límite', async () => {
  const c = faro.init({ ...commonOpts(1), maxQueueSize: 5 });
  try {
    for (let i = 0; i < 50; i++) c.log({ level: 'INFO', message: `e${i}` });
    assert.equal(c.queue.length, 5, 'la cola debe pararse en el cap');
  } finally {
    await c.close();
  }
});

// ---- 2. retry sobre 5xx ----

test('5xx: el batch se re-encola', async () => {
  let calls = 0;
  const { server, port } = await startServer((_req, res) => {
    calls += 1;
    if (calls === 1) {
      res.writeHead(503);
      res.end('caído');
    } else {
      res.writeHead(200);
      res.end('{}');
    }
  });

  const c = faro.init(commonOpts(port));
  try {
    c.log({ level: 'INFO', message: 'reintentar-me' });
    await c.flush(); // 503 → re-encola
    await c.flush(); // ahora OK
    await delay(50);
    assert.ok(calls >= 2, 'segundo intento debe haber llegado');
  } finally {
    await c.close();
    server.close();
  }
});

// ---- 3. beforeSend ----

test('beforeSend: null descarta', async () => {
  const seen = [];
  const { server, port } = await startServer((_req, res, body) => {
    seen.push(JSON.parse(body));
    res.writeHead(200);
    res.end('{}');
  });

  const c = faro.init({
    ...commonOpts(port),
    beforeSend: (e) => (e.message.includes('descartar') ? null : e),
  });
  try {
    c.log({ level: 'INFO', message: 'guardar' });
    c.log({ level: 'INFO', message: 'descartar' });
    c.log({ level: 'INFO', message: 'guardar también' });
    await c.flush();
    await delay(50);
    const msgs = seen[0].logs.map((l) => l.message);
    assert.deepEqual(msgs, ['guardar', 'guardar también']);
  } finally {
    await c.close();
    server.close();
  }
});

// ---- 4. scrubbing ----

test('scrubbing: scrubFields + scrubPatterns', async () => {
  const seen = [];
  const { server, port } = await startServer((_req, res, body) => {
    seen.push(JSON.parse(body));
    res.writeHead(200);
    res.end('{}');
  });

  const c = faro.init(commonOpts(port));
  try {
    c.log({
      level: 'INFO',
      message: 'auth con eyJabc.def.ghi y key sk-abcdefghijklmnop',
      attributes: {
        'user.password': 'p4ss',
        'http.request.header.authorization': 'Bearer x',
        'safe.field': 'visible',
        'embedded': 'ghp_aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa',
      },
    });
    await c.flush();
    await delay(50);
    const log = seen[0].logs[0];
    assert.equal(log.attributes['user.password'], '[REDACTED]');
    assert.equal(log.attributes['http.request.header.authorization'], '[REDACTED]');
    assert.equal(log.attributes['safe.field'], 'visible');
    assert.equal(log.attributes['embedded'], '[REDACTED]');
    assert.ok(!log.message.includes('eyJabc'), 'JWT redactado en message');
    assert.ok(!log.message.includes('sk-abcdef'), 'sk-* redactado en message');
  } finally {
    await c.close();
    server.close();
  }
});

// ---- Product events API: track / identify / screen / alias ----

test('track: envía evento mobile a /api/v1/ingest/events', async () => {
  const seen = [];
  const { server, port } = await startServer((req, res, body) => {
    seen.push({ url: req.url, body: JSON.parse(body) });
    res.writeHead(200);
    res.end('{}');
  });

  const c = faro.init(commonOpts(port));
  try {
    c.track('checkout_completed', { amount: 99.5, currency: 'USD' });
    await c.flush();
    await delay(50);

    const eventBatch = seen.find((s) => s.url.endsWith('/api/v1/ingest/events'));
    assert.ok(eventBatch, `esperaba POST a /events; visto: ${JSON.stringify(seen.map((s) => s.url))}`);
    assert.equal(eventBatch.body.service, 'expo-tests');
    const event = eventBatch.body.events[0];
    assert.equal(event.type, 'track');
    assert.equal(event.name, 'checkout_completed');
    assert.deepEqual(event.properties, { amount: 99.5, currency: 'USD' });
    assert.ok(event.distinct_id.startsWith('anon_'));
    assert.equal(event.distinct_id, event.anonymous_id);
    assert.equal(event.session_id, '');
    assert.equal(event.source, 'mobile');
  } finally {
    await c.close();
    server.close();
  }
});

test('identify: fija distinct_id para eventos siguientes', async () => {
  const seen = [];
  const { server, port } = await startServer((_req, res, body) => {
    seen.push(JSON.parse(body));
    res.writeHead(200);
    res.end('{}');
  });

  const c = faro.init(commonOpts(port));
  try {
    c.identify('user_42', { email: 'a@b.com', plan: 'pro' });
    c.track('after_login');
    await c.flush();
    await delay(50);

    const events = seen.flatMap((b) => b.events ?? []);
    const identify = events.find((e) => e.type === 'identify');
    const track = events.find((e) => e.type === 'track');
    assert.ok(identify, 'debe llegar un identify');
    assert.equal(identify.name, '$identify');
    assert.equal(identify.distinct_id, 'user_42');
    assert.deepEqual(identify.user_properties, { email: 'a@b.com', plan: 'pro' });
    assert.ok(track, 'debe llegar el track posterior');
    assert.equal(track.distinct_id, 'user_42');
  } finally {
    await c.close();
    server.close();
  }
});

test('screen: emite screen view mobile con propiedades', async () => {
  const seen = [];
  const { server, port } = await startServer((_req, res, body) => {
    seen.push(JSON.parse(body));
    res.writeHead(200);
    res.end('{}');
  });

  const c = faro.init(commonOpts(port));
  try {
    c.screen('CheckoutSuccess', { source: 'cart' });
    await c.flush();
    await delay(50);

    const event = seen.flatMap((b) => b.events ?? []).find((e) => e.type === 'screen');
    assert.ok(event, 'debe llegar un screen event');
    assert.equal(event.name, 'CheckoutSuccess');
    assert.deepEqual(event.properties, { source: 'cart' });
    assert.equal(event.source, 'mobile');
  } finally {
    await c.close();
    server.close();
  }
});

test('alias: fusiona anonymous_id previo con user post-login', async () => {
  const seen = [];
  const { server, port } = await startServer((_req, res, body) => {
    seen.push(JSON.parse(body));
    res.writeHead(200);
    res.end('{}');
  });

  const c = faro.init(commonOpts(port));
  try {
    c.alias('anonymous_abc123', 'user_42');
    c.track('post_alias');
    await c.flush();
    await delay(50);

    const events = seen.flatMap((b) => b.events ?? []);
    const alias = events.find((e) => e.type === 'alias');
    const track = events.find((e) => e.type === 'track');
    assert.ok(alias, 'debe llegar un alias');
    assert.equal(alias.name, '$alias');
    assert.equal(alias.anonymous_id, 'anonymous_abc123');
    assert.equal(alias.distinct_id, 'user_42');
    assert.equal(track.distinct_id, 'user_42');
  } finally {
    await c.close();
    server.close();
  }
});
